src/Features/Core/Portable/Formatting/AbstractFormattingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal class AbstractFormattingDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public AbstractFormattingDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.FormattingDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_document), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_document), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeSyntaxTree);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            if (!(context.Options is WorkspaceAnalyzerOptions workspaceAnalyzerOptions))
+            {
+                return;
+            }
+
+            var options = context.Options.GetDocumentOptionSetAsync(context.Tree, context.CancellationToken).GetAwaiter().GetResult();
+            if (options == null)
+            {
+                return;
+            }
+
+            var workspace = workspaceAnalyzerOptions.Services.Workspace;
+            var formattingChanges = Formatter.GetFormattedTextChanges(context.Tree.GetRoot(context.CancellationToken), workspace, options, context.CancellationToken);
+            foreach (var change in formattingChanges)
+            {
+                var span = change.Span;
+                if (span.IsEmpty)
+                {
+                    // Need to add characters
+                    span = new TextSpan(change.Span.Start, 1);
```

Comment:
you probably want to double check. diagnostic span == 0 might not shown but tracking span being 0 might be shown (by extending it to 1 length). since we do not remove diagnostics even if text is deleted from editor.

Reply:
okay. I think we sometime remove and some time not. I think we don't remove in venus/razor case... can't remember exact case on top of my head.

Reply:
> you probably want to double check

it's definitely supported.  When the compiler reports a missing token, they have a 0-width token and they just report the diagnostic on that missing token.  There is no adjustment from them (or anything higher in the stack) that ensures the diagnostic itself has non-zero length.

The only thing that is ever done about this is that the squiggle adornment is ensured to have length of at least 1 so it will actually be visible in the UI.